From: Heather Lemon <heather.lemon@canonical.com>
Date: Wed, 19 May 2021 08:19:03 -0600
Subject: [PATCH] libfreeipmi/ipmi-sensor-read.c: Incorrect sensor
reading reported. The ipmi-sensors command
reads the wrong temperature sensor on a Lenovo BMC.
When the LUN number is not the default (00b) the sensor is reporting
back an incorrect temperature reading.
You are allowed to change the default, but unless otherwise specified,
commands listed as mandatory must be accessed with LUN 00b.
Note ipmi-tools correctly reports the sensor temperature
because it knows how to use the alternative LUN number correctly.
Forwarded: no
Last-Updated: 2021-05-20
Origin: upstream, https://github.com/chu11/freeipmi-mirror/commit/82db0fb1d834ff404444483304064486f1481610
Bug-Ubuntu: https://bugs.launchpad.net/bugs/1926299

---
 ChangeLog                                  | 8 ++++++++
 libfreeipmi/sensor-read/ipmi-sensor-read.c | 3 ++-
 2 files changed, 10 insertions(+), 1 deletion(-)

Index: freeipmi-1.6.6/ChangeLog
===================================================================
--- freeipmi-1.6.6.orig/ChangeLog
+++ freeipmi-1.6.6/ChangeLog
@@ -1,3 +1,11 @@
+2021-05-29 Heather Lemon <heather.lemon@canonical.com>
+
+	* libfreeipmi/sensor-read/ipmi-sensor-read.c: (LP#1926299)
+	Incorrect sensor reading reported.
+	BMC's that use a non-zero 'LUN' number (00b-10b)
+	will display an incorrect sensor temperature reading unless
+	otherwise specified via the ipmiV2.0 spec and must be accessed via LUN 00b.
+
 2020-09-03 Christian Ehrhardt <christian.ehrhardt@canonical.com>

 	* libfreeipmi/locate/ipmi-locate-acpi-spmi.c
Index: freeipmi-1.6.6/libfreeipmi/sensor-read/ipmi-sensor-read.c
===================================================================
--- freeipmi-1.6.6.orig/libfreeipmi/sensor-read/ipmi-sensor-read.c
+++ freeipmi-1.6.6/libfreeipmi/sensor-read/ipmi-sensor-read.c
@@ -40,6 +40,7 @@
 #include "freeipmi/record-format/ipmi-sdr-record-format.h"
 #include "freeipmi/spec/ipmi-channel-spec.h"
 #include "freeipmi/spec/ipmi-comp-code-spec.h"
+#include "freeipmi/spec/ipmi-ipmb-lun-spec.h"
 #include "freeipmi/spec/ipmi-slave-address-spec.h"
 #include "freeipmi/spec/ipmi-sensor-units-spec.h"
 #include "freeipmi/util/ipmi-sensor-and-event-code-tables-util.h"
@@ -563,7 +564,7 @@ ipmi_sensor_read (ipmi_sensor_read_ctx_t
    */
   if (!(ctx->flags & IPMI_SENSOR_READ_FLAGS_ASSUME_BMC_OWNER))
     {
-      if (slave_address == IPMI_SLAVE_ADDRESS_BMC)
+      if (slave_address == IPMI_SLAVE_ADDRESS_BMC && sensor_owner_lun == IPMI_BMC_IPMB_LUN_BMC)
         {
           if (_get_sensor_reading (ctx,
                                    sensor_number,
